### PR TITLE
Make fishtest server python3 compatible

### PR DESCRIPTION
**Pull request by tomtor**
Also very few changes required.

Tested on Python2 AND Python3

### CODE REVIEW

**Comment by tomtor** [comment]
No hurry to merge this, although it should do no harm either... 
Perhaps it could be combined with an future infrastructure update.

File: fishtest/fishtest/rundb.py [hunk 1]
```diff
@@ -333,7 +333,7 @@ def active_run_lock(self, id):
       self.purge_count = self.purge_count + 1
       if self.purge_count > 100000:
         old = time.time() - 10000
-        self.active_runs = dict((k,v) for k, v in self.active_runs.iteritems() if v['time'] >= old)
+        self.active_runs = dict((k,v) for k, v in list(self.active_runs.items()) if v['time'] >= old)
```
Comment (tomtor) [inline fishtest/fishtest/rundb.py]: I do not think the list () is needed in this context, but it was advised by the 2to3 tool. This code is not performance critical so it does not really matter.

**Comment by Stefano80** [comment]
It looks like a cool endeavor, @tomtor , thank you.

**Comment by Stefano80** [comment]
Hi @tomtor , is this done?

**Comment by tomtor** [comment]
Yes, it is, but deploying it with Python3 makes most sense on a refreshed OS.

I have only tested it with Python 2.7 not with 2.6 so I would not advise to use it on the current production server configuration.

**Comment by Stefano80** [comment]
Ok, then let us wait for more convenient times.

**Comment by tomtor** [comment]
Note scavenge.py has a print that must be fixed!

**Comment by tomtor** [comment]
@ppigazzini Just for reference, I installed pypy 3.6 from https://github.com/squeaky-pl/portable-pypy#portable-pypy-distribution-for-linux and this branch combined with https://github.com/glinscott/fishtest/pull/43 started fine on my server.

Just two notes, in the `virtualenv-pypy` :
* You must install fortran before installing numpy: `sudo apt install gfortran`
* You must install numpy: `pip install numpy` before executing `python setup.py install`

File: fishtest/fishtest/views.py [hunk 2]
```diff
@@ -752,7 +752,7 @@ def get_chi2(tasks, bad_users):
   if len(users) == 0:
     return results
 
-  observed = numpy.array(users.values())
+  observed = numpy.array(list(users.values()))
```
Comment (tomtor) [inline fishtest/fishtest/views.py]: @ppigazzini Had to change this after the recent merge of Numpy code

File: fishtest/fishtest/views.py [hunk 3]
```diff
@@ -784,7 +784,7 @@ def get_chi2(tasks, bad_users):
   std_error = numpy.sqrt(expected * numpy.outer((1 - row_sums / grand_total), (1 - column_sums / grand_total)))
   adj_residual = raw_residual / std_error
   for idx in range(len(users)):
-    users[users.keys()[idx]] = numpy.max(numpy.abs(adj_residual[idx]))
+    users[list(users.keys())[idx]] = numpy.max(numpy.abs(adj_residual[idx]))
```
Comment (tomtor) [inline fishtest/fishtest/views.py]: @ppigazzini Idem

**Comment by tomtor** [comment]
https://pythonclock.org/

**Comment by ppigazzini** [comment]
@tomtor it doesn't work for me, I'm reading all zeros...

:P

**Comment by ppigazzini** [comment]
@tomtor DEV running on python3

**Comment by ppigazzini** [comment]
@tomtor mmm, my worker shows a problem

```
Exception accessing host:
Traceback (most recent call last):
  File "worker.py", line 109, in worker
    req = json.loads(req.text)
  File "/usr/lib/python3.6/json/__init__.py", line 354, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python3.6/json/decoder.py", line 339, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python3.6/json/decoder.py", line 357, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```

**Comment by ppigazzini** [comment]
@tomtor `sudo journalctl -u fishtest@6543` :

```
Jan 02 17:29:36 vps1572612706 pserve[6275]: 2020-01-02 17:29:36,205 ERROR [waitress][waitress] Exception while serving /api/request_task
Jan 02 17:29:36 vps1572612706 pserve[6275]: Traceback (most recent call last):
Jan 02 17:29:36 vps1572612706 pserve[6275]:   File "/home/usr00/fishtest/fishtest/env/lib/python3.6/site-packages/pyramid/tweens.py", line 13, in _error_handler
Jan 02 17:29:36 vps1572612706 pserve[6275]:     response = request.invoke_exception_view(exc_info)
Jan 02 17:29:36 vps1572612706 pserve[6275]:   File "/home/usr00/fishtest/fishtest/env/lib/python3.6/site-packages/pyramid/view.py", line 779, in invoke_exception_view
Jan 02 17:29:36 vps1572612706 pserve[6275]:     raise HTTPNotFound
Jan 02 17:29:36 vps1572612706 pserve[6275]: pyramid.httpexceptions.HTTPNotFound: The resource could not be found.
Jan 02 17:29:36 vps1572612706 pserve[6275]: During handling of the above exception, another exception occurred:
Jan 02 17:29:36 vps1572612706 pserve[6275]: Traceback (most recent call last):
Jan 02 17:29:36 vps1572612706 pserve[6275]:   File "/home/usr00/fishtest/fishtest/env/lib/python3.6/site-packages/waitress/channel.py", line 349, in service
Jan 02 17:29:36 vps1572612706 pserve[6275]:     task.service()
Jan 02 17:29:36 vps1572612706 pserve[6275]:   File "/home/usr00/fishtest/fishtest/env/lib/python3.6/site-packages/waitress/task.py", line 169, in service
Jan 02 17:29:36 vps1572612706 pserve[6275]:     self.execute()
Jan 02 17:29:36 vps1572612706 pserve[6275]:   File "/home/usr00/fishtest/fishtest/env/lib/python3.6/site-packages/waitress/task.py", line 439, in execute
Jan 02 17:29:36 vps1572612706 pserve[6275]:     app_iter = self.channel.server.application(environ, start_response)
Jan 02 17:29:36 vps1572612706 pserve[6275]:   File "/home/usr00/fishtest/fishtest/env/lib/python3.6/site-packages/waitress/proxy_headers.py", line 65, in translate_proxy_headers
Jan 02 17:29:36 vps1572612706 pserve[6275]:     return app(environ, start_response)
Jan 02 17:29:36 vps1572612706 pserve[6275]:   File "/home/usr00/fishtest/fishtest/env/lib/python3.6/site-packages/pyramid/router.py", line 270, in __call__
Jan 02 17:29:36 vps1572612706 pserve[6275]:     response = self.execution_policy(environ, self)
Jan 02 17:29:36 vps1572612706 pserve[6275]:   File "/home/usr00/fishtest/fishtest/env/lib/python3.6/site-packages/pyramid/router.py", line 279, in default_execution_policy
Jan 02 17:29:36 vps1572612706 pserve[6275]:     return request.invoke_exception_view(reraise=True)
Jan 02 17:29:36 vps1572612706 pserve[6275]:   File "/home/usr00/fishtest/fishtest/env/lib/python3.6/site-packages/pyramid/view.py", line 778, in invoke_exception_view
Jan 02 17:29:36 vps1572612706 pserve[6275]:     reraise_(*exc_info)
Jan 02 17:29:36 vps1572612706 pserve[6275]:   File "/home/usr00/fishtest/fishtest/env/lib/python3.6/site-packages/pyramid/compat.py", line 179, in reraise
Jan 02 17:29:36 vps1572612706 pserve[6275]:     raise value
Jan 02 17:29:36 vps1572612706 pserve[6275]:   File "/home/usr00/fishtest/fishtest/env/lib/python3.6/site-packages/pyramid/router.py", line 277, in default_execution_policy
Jan 02 17:29:36 vps1572612706 pserve[6275]:     return router.invoke_request(request)
Jan 02 17:29:36 vps1572612706 pserve[6275]:   File "/home/usr00/fishtest/fishtest/env/lib/python3.6/site-packages/pyramid/router.py", line 249, in invoke_request
Jan 02 17:29:36 vps1572612706 pserve[6275]:     response = handle_request(request)
Jan 02 17:29:36 vps1572612706 pserve[6275]:   File "/home/usr00/fishtest/fishtest/env/lib/python3.6/site-packages/pyramid/tweens.py", line 43, in excview_tween
Jan 02 17:29:36 vps1572612706 pserve[6275]:     response = _error_handler(request, exc)
Jan 02 17:29:36 vps1572612706 pserve[6275]:   File "/home/usr00/fishtest/fishtest/env/lib/python3.6/site-packages/pyramid/tweens.py", line 17, in _error_handler
Jan 02 17:29:36 vps1572612706 pserve[6275]:     reraise(*exc_info)
Jan 02 17:29:36 vps1572612706 pserve[6275]:   File "/home/usr00/fishtest/fishtest/env/lib/python3.6/site-packages/pyramid/compat.py", line 179, in reraise
Jan 02 17:29:36 vps1572612706 pserve[6275]:     raise value
Jan 02 17:29:36 vps1572612706 pserve[6275]:   File "/home/usr00/fishtest/fishtest/env/lib/python3.6/site-packages/pyramid/tweens.py", line 41, in excview_tween
Jan 02 17:29:36 vps1572612706 pserve[6275]:     response = handler(request)
Jan 02 17:29:36 vps1572612706 pserve[6275]:   File "/home/usr00/fishtest/fishtest/env/lib/python3.6/site-packages/pyramid/router.py", line 148, in handle_request
Jan 02 17:29:36 vps1572612706 pserve[6275]:     registry, request, context, context_iface, view_name
Jan 02 17:29:36 vps1572612706 pserve[6275]:   File "/home/usr00/fishtest/fishtest/env/lib/python3.6/site-packages/pyramid/view.py", line 667, in _call_view
Jan 02 17:29:36 vps1572612706 pserve[6275]:     response = view_callable(context, request)
Jan 02 17:29:36 vps1572612706 pserve[6275]:   File "/home/usr00/fishtest/fishtest/env/lib/python3.6/site-packages/pyramid/viewderivers.py", line 436, in rendered_view
Jan 02 17:29:36 vps1572612706 pserve[6275]:     result = view(context, request)
Jan 02 17:29:36 vps1572612706 pserve[6275]:   File "/home/usr00/fishtest/fishtest/env/lib/python3.6/site-packages/pyramid/viewderivers.py", line 144, in _requestonly_view
Jan 02 17:29:36 vps1572612706 pserve[6275]:     response = view(request)
Jan 02 17:29:36 vps1572612706 pserve[6275]:   File "/home/usr00/fishtest/fishtest/fishtest/api.py", line 93, in request_task
Jan 02 17:29:36 vps1572612706 pserve[6275]:     result = request.rundb.request_task(worker_info)
Jan 02 17:29:36 vps1572612706 pserve[6275]:   File "/home/usr00/fishtest/fishtest/fishtest/rundb.py", line 321, in request_task
Jan 02 17:29:36 vps1572612706 pserve[6275]:     return self.sync_request_task(worker_info)
Jan 02 17:29:36 vps1572612706 pserve[6275]:   File "/home/usr00/fishtest/fishtest/fishtest/rundb.py", line 334, in sync_request_task
Jan 02 17:29:36 vps1572612706 pserve[6275]:     self.calc_itp(run)
Jan 02 17:29:36 vps1572612706 pserve[6275]:   File "/home/usr00/fishtest/fishtest/fishtest/rundb.py", line 297, in calc_itp
Jan 02 17:29:36 vps1572612706 pserve[6275]:     run['results_info'] = format_results(results, run)
Jan 02 17:29:36 vps1572612706 pserve[6275]:   File "/home/usr00/fishtest/fishtest/fishtest/views.py", line 675, in format_results
Jan 02 17:29:36 vps1572612706 pserve[6275]:     elo_model=elo_model
Jan 02 17:29:36 vps1572612706 pserve[6275]:   File "/home/usr00/fishtest/fishtest/fishtest/stats/stat_util.py", line 213, in SPRT
Jan 02 17:29:36 vps1572612706 pserve[6275]:     LLR_,overshoot=LLRcalc.LLR_logistic(lelo0,lelo1,R['pentanomial'])
Jan 02 17:29:36 vps1572612706 pserve[6275]:   File "/home/usr00/fishtest/fishtest/fishtest/stats/LLRcalc.py", line 168, in LLR_logistic
Jan 02 17:29:36 vps1572612706 pserve[6275]:     return N*LLR(pdf,s0,s1),overshoot
Jan 02 17:29:36 vps1572612706 pserve[6275]:   File "/home/usr00/fishtest/fishtest/fishtest/stats/LLRcalc.py", line 70, in LLR
Jan 02 17:29:36 vps1572612706 pserve[6275]:     return stats(LLRjumps(pdf,s0,s1))[0]
Jan 02 17:29:36 vps1572612706 pserve[6275]:   File "/home/usr00/fishtest/fishtest/fishtest/stats/LLRcalc.py", line 60, in LLRjumps
Jan 02 17:29:36 vps1572612706 pserve[6275]:     pdf0,pdf1=[MLE(pdf,s) for s in (s0,s1)]
Jan 02 17:29:36 vps1572612706 pserve[6275]:   File "/home/usr00/fishtest/fishtest/fishtest/stats/LLRcalc.py", line 60, in <listcomp>
Jan 02 17:29:36 vps1572612706 pserve[6275]:     pdf0,pdf1=[MLE(pdf,s) for s in (s0,s1)]
Jan 02 17:29:36 vps1572612706 pserve[6275]:   File "/home/usr00/fishtest/fishtest/fishtest/stats/LLRcalc.py", line 29, in MLE
Jan 02 17:29:36 vps1572612706 pserve[6275]:     res=brentq.brentq(f,l+epsilon,u-epsilon,epsilon=epsilon)
Jan 02 17:29:36 vps1572612706 pserve[6275]: NameError: name 'brentq' is not defined
```

**Comment by tomtor** [comment]
@ppigazzini Hope this is the last wrong import...

**Comment by ppigazzini** [comment]
@tomtor the workers are working now :)

**Comment by ppigazzini** [comment]
@tomtor I got this error testing the python3 server setup script

```
Created symlink /etc/systemd/system/multi-user.target.wants/mongod.service → /lib/systemd/system/mongod.service.
Created symlink /etc/systemd/system/multi-user.target.wants/fishtest@6543.service → /etc/systemd/system/fishtest@.service.
Created symlink /etc/systemd/system/multi-user.target.wants/fishtest@6544.service → /etc/systemd/system/fishtest@.service.
Traceback (most recent call last):
  File "<stdin>", line 3, in <module>
  File "/home/fishtest/fishtest/fishtest/fishtest/__init__.py", line 8, in <module>
    from rundb import RunDb
  File "/home/fishtest/fishtest/fishtest/fishtest/rundb.py", line 16, in <module>
    from views import format_results
  File "/home/fishtest/fishtest/fishtest/fishtest/views.py", line 20, in <module>
    import fishtest.stats.stat_util
  File "/home/fishtest/fishtest/fishtest/fishtest/stats/stat_util.py", line 3, in <module>
    import LLRcalc,sprt
ModuleNotFoundError: No module named 'LLRcalc'
no crontab for fishtest
connect a browser to:
http://192.168.84.101
```

If I recall correctly we could relax the requirements on numpy (<1.16) and scipy (<1.2).

**Comment by tomtor** [comment]
@ppigazzini `stat_util.py` in this PR reads:

```
from __future__ import division
import math,copy
from fishtest.stats import LLRcalc
from fishtest.stats import sprt
```

So you are installing files from a branch without this PR merged.

I forgot about the version requirements, thanks for the hint!

**Comment by ppigazzini** [comment]
@tomtor the scheduled jobs run fine on DEV but I got this error with a scan_all:

```
${VENV}/bin/python ${UPATH}/delta_update_users.py scan_all
scan all
Traceback (most recent call last):
  File "/home/usr00/fishtest/fishtest/utils/delta_update_users.py", line 43, in process_run
    info[username]['last_updated'] = max(task['last_updated'], info[username]['last_updated'])
KeyError: 'jobbautista9'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/usr00/fishtest/fishtest/utils/delta_update_users.py", line 173, in <module>
    main()
  File "/home/usr00/fishtest/fishtest/utils/delta_update_users.py", line 170, in main
    update_users()
  File "/home/usr00/fishtest/fishtest/utils/delta_update_users.py", line 118, in update_users
    process_run(run, info, deltas)
  File "/home/usr00/fishtest/fishtest/utils/delta_update_users.py", line 45, in process_run
    info[username]['last_updated'] = task['last_updated']
KeyError: 'jobbautista9'
```

ps: server setup script successfully updated :)

**Comment by ppigazzini** [comment]
@tomtor I fear that boto2 (used for backup on s3) doesn't work with python 3.6.9
```
${VENV}/bin/fetch_file s3://fishtest/backup/archive/20200101/dump.tar.gz -o dump.tar.gz
Traceback (most recent call last):
  File "/home/usr00/fishtest/fishtest/env/bin/fetch_file", line 44, in <module>
    open(options.outfile, "w").write(f.read())
TypeError: write() argument must be str, not bytes
```

https://github.com/boto/boto

**Comment by tomtor** [comment]
@ppigazzini Regarding boto2. In Python3 binary files must be opened with "wb" or "rb". 

Can you adapt/test the script?

**Comment by tomtor** [comment]
See also https://boto3.amazonaws.com/v1/documentation/api/latest/guide/migration.html

**Comment by ppigazzini** [comment]
@tomtor another error running scan_all
```
$ python3 delta_update_users.py scan_all
scan all
not in info:  jobbautista9
Warning: skipping repeated run!
not in info:  Akusari
not in info:  Akusari
Traceback (most recent call last):
  File "delta_update_users.py", line 176, in <module>
    main()
  File "delta_update_users.py", line 173, in main
    update_users()
  File "delta_update_users.py", line 121, in update_users
    process_run(run, info, deltas)
  File "delta_update_users.py", line 26, in process_run
    info[username]['tests'] += 1
KeyError: 'tchoty'
```

**Comment by ppigazzini** [comment]
@tomtor I foresee that this will be the last error :) 
```
$ python3 delta_update_users.                                                                                                                               py scan_all
scan all
not in info:  jobbautista9
Warning: skipping repeated run!
not in info:  Akusari
not in info:  Akusari
not in info:  tchoty
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  tmvoss
not in info:  Kobal
not in info:  skiminki
Traceback (most recent call last):
  File "delta_update_users.py", line 179, in <module>
    main()
  File "delta_update_users.py", line 176, in main
    update_users()
  File "delta_update_users.py", line 124, in update_users
    process_run(run, info, deltas)
  File "delta_update_users.py", line 31, in process_run
    tc = parse_tc(run['args']['tc'])
  File "/home/usr00/fishtest/fishtest/fishtest/views.py", line 167, in parse_tc
    increment = float(chunks[1])
ValueError: could not convert string to float: '0.0.5'
```

**Comment by tomtor** [comment]
I expect this garbage is also in the prod database?

**Comment by ppigazzini** [comment]
@tomtor unfortunately I think so...

File: fishtest/setup.py [hunk 4]
```diff
@@ -11,10 +11,10 @@
     'pyramid_mako',
     'waitress',
     'pymongo',
-    'numpy<1.16',
-    'scipy<1.2',
+    'numpy',
+    'scipy',
     'requests',
-    'boto'
+    'boto',
```
Comment (ppigazzini) [inline fishtest/setup.py]: @tomtor I deleted boto package in #486, already added the final comma.

**Comment by ppigazzini** [comment]
@tomtor I loaded on DEV a fresh copy of the DB, I confirm the presence of the garbage :(
Update:
- your last commit doesn't fix the latest error
- the garbage breaks the scan_all also with python2
- I installed on PROD python3 python3-pip python3-venv 
- I written on PROD the update scripts:
  - for python3 (`x/update_fishtest_p3.sh`), this can also merge the PR if we want wait before merging the PR on github (the restore to python2 will be a little more fast)
  - for python2 (`x/update_fishtest_p2.sh`) to revert in case of problems

I'm for the python3 switch asap in order to have the weekend for debug, let me know your preference (when switch and the type of merge [on github or on PROD with the script]).

**Comment by tomtor** [comment]
I suggest switching on prod with the script.

I am away the rest of the evening, cannot assist until tomorrow morning, but please proceed when you are confident. Will look at the scan all problem later. Can you paste the output from dev?

**Comment by tomtor** [comment]
@ppigazzini Pushed one more fix for "scan all"

**Comment by ppigazzini** [comment]
@tomtor w/ the commit of 8 hour ago the full scan had the same output of this comment https://github.com/glinscott/fishtest/pull/276#issuecomment-570548031

This is the full scan output with latest commit:
```
$ python3 delta_update_users.py scan_all
scan all
not in info:  jobbautista9
Warning: skipping repeated run!
Warning: skipping repeated run!
Warning: skipping repeated run!
not in info:  Akusari
not in info:  Akusari
not in info:  tchoty
Warning: skipping repeated run!
Warning: skipping repeated run!
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  tmvoss
not in info:  Kobal
not in info:  skiminki
Exception on run:  {'_id': ObjectId('5c3daa070ebc596a450cffd4'), 'results_info': {'info': ['Pending...'], 'style': ''}, 'tasks': [{'worker_key': 'ChessDBCN-8cores-79659', 'num_games': 250, 'worker_info': {'username': 'xoto', 'min_threads': 1, 'unique_key': '4d60ba0f-1278-4077-b028-ed68c62a18c2', 'remote_addr': '212.159.20.170', 'max_memory': 12041, 'uname': 'Linux 4.13.0-38-generic', 'version': '67:2', 'architecture': ['64bit', ''], 'concurrency': 12, 'country_code': 'GB'}, 'residual': 0.0, 'residual_color': '#44EB44', 'active': False, 'last_updated': datetime.datetime(2019, 1, 16, 17, 9, 50, 209000), 'pending': False}, {'worker_key': 'gvreuls-15cores-202', 'num_games': 250, 'worker_info': {'username': 'jromang', 'min_threads': 1, 'unique_key': '95c5ee30-f96d-4985-8351-0277f73848a5', 'remote_addr': '109.13.128.89', 'max_memory': 16049, 'uname': 'Linux 4.20.0-1-MANJARO', 'version': '67:3', 'architecture': ['64bit', ''], 'concurrency': 15}, 'residual': 0.0, 'residual_color': '#44EB44', 'active': False, 'last_updated': datetime.datetime(2019, 1, 16, 17, 9, 53, 285000), 'pending': False}, {'worker_key': 'xoto-12cores-1', 'num_games': 250, 'worker_info': {'username': 'gvreuls', 'min_threads': 1, 'unique_key': '94219f44-b76f-4dfb-89c1-45171614be66', 'remote_addr': '80.100.156.171', 'max_memory': 15922, 'uname': 'Linux 4.19.14-300.fc29.x86_64', 'version': '67:2', 'architecture': ['64bit', ''], 'concurrency': 15, 'country_code': 'NL'}, 'residual': 0.0, 'residual_color': '#44EB44', 'active': False, 'last_updated': datetime.datetime(2019, 1, 16, 17, 9, 54, 529000), 'pending': False}], 'last_updated': datetime.datetime(2019, 1, 15, 9, 38, 15, 450000), 'deleted': True, 'results_stale': False, 'start_time': datetime.datetime(2019, 1, 15, 9, 38, 15, 450000), 'args': {'book_depth': '8', 'internal_priority': -1547599095.0, 'new_signature': '3739723', 'auto_purge': True, 'tests_repo': 'https://github.com/xoto10/stockfish-xoto10', 'new_tag': 'ttvalue1', 'sprt': {'alpha': 0.05, 'beta': 0.05, 'drawelo': 240.0, 'elo0': 0.5, 'elo1': 4.5}, 'priority': 0, 'base_signature': '3739723', 'book': '2moves_v1.pgn', 'tc': '5+0.0.5', 'username': 'xoto', 'base_tag': 'master', 'msg_base': 'Remove pvExact', 'new_options': 'Hash=8', 'resolved_new': 'caa80dbcfb3b369e4c6c999688a028d62f68aa82', 'threads': 8, 'resolved_base': '5446e6f408f2ed7fa281dbe0097c46674d193260', 'base_options': 'Hash=8', 'info': 'use ttvalue is set and eval not set and pvhit.', 'num_games': 250000, 'msg_new': 'use ttvalue is set and eval not set and pvhit. bench 3739723', 'throughput': 250}, 'results': {'wins': 0, 'losses': 0, 'draws': 0}, 'approver': '31m059', 'finished': True, 'approved': True}
not in info:  anteres2017
not in info:  TheIncarnadineZebra
not in info:  TheIncarnadineZebra
not in info:  e4!!
not in info:  e4!!
not in info:  e4!!
not in info:  e4!!
not in info:  HenryElliott
not in info:  Write Essays Online
not in info:  School Assignments
not in info:  My Paper Writer
not in info:  asvega
not in info:  xcrTTErR
not in info:  xcrTTErR
not in info:  Aa Seat Assignment
not in info:  Essay-writer.Org
not in info:  Write Essay Service
not in info:  Shadousky
not in info:  Buy Custom Essays
not in info:  pmalcolm
not in info:  Exception on run:  Traceback (most recent call last):
  File "delta_update_users.py", line 129, in update_users
    process_run(run, info, deltas)
  File "delta_update_users.py", line 27, in process_run
    print('not in info: ', username)
UnicodeEncodeError: 'ascii' codec can't encode character '\xfb' in position 5: ordinal not in range(128)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "delta_update_users.py", line 189, in <module>
    main()
  File "delta_update_users.py", line 186, in main
    update_users()
  File "delta_update_users.py", line 131, in update_users
    print("Exception on run: ", run)
UnicodeEncodeError: 'ascii' codec can't encode character '\xfb' in position 7747: ordinal not in range(128)
```

**Comment by ppigazzini** [comment]
@tomtor PROD on python 3.6.9 at 22:56:04 CET, finger crossed :)

**Comment by ppigazzini** [comment]
pserve and mongod logs are clean, the scheduled jobs are running fine, CPU and memory load at the moment are low.

**Comment by ppigazzini** [comment]
@tomtor scan_all started, it will take some time.

**Comment by ppigazzini** [comment]
@tomtor scan all worked fine !

```
$ python delta_update_users.py scan_all
scan all
not in info:  jobbautista9
Warning: skipping repeated run!
Warning: skipping repeated run!
Warning: skipping repeated run!
Warning: skipping repeated run!
not in info:  Akusari
not in info:  Akusari
not in info:  tchoty
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  Ndestroyer5237
not in info:  tmvoss
not in info:  Kobal
not in info:  skiminki
Exception on run:  5c3daa070ebc596a450cffd4
not in info:  anteres2017
not in info:  TheIncarnadineZebra
not in info:  TheIncarnadineZebra
not in info:  e4!!
not in info:  e4!!
not in info:  e4!!
not in info:  e4!!
not in info:  HenryElliott
not in info:  Write Essays Online
not in info:  School Assignments
not in info:  My Paper Writer
not in info:  asvega
not in info:  xcrTTErR
not in info:  xcrTTErR
not in info:  Aa Seat Assignment
not in info:  Essay-writer.Org
not in info:  Write Essay Service
not in info:  Shadousky
not in info:  Buy Custom Essays
not in info:  pmalcolm
not in info:  Exception on run:  5a4be5bc0ebc590ccbb8c5ad
not in info:  WireInTheShark
not in info:  FeCheDeU15
not in info:  FeCheDeU15
not in info:  Simpo137
not in info:  Simpo137
not in info:  lithandi
not in info:  Simpo137
not in info:  buddy_dacote
not in info:  AlexZet
not in info:  SweD
not in info:  SweD
not in info:  violinandchess1
not in info:  janostry
not in info:  retrovar
not in info:  hawkesid
not in info:  Asylzatwww
not in info:  adam200k
not in info:  andrewzhernov
not in info:  danazol
not in info:  marshmallow
not in info:  geo17
not in info:  mchvs
not in info:  mchvs
not in info:  mchvs
not in info:  Fytch
not in info:  ennsharma
not in info:  ennsharma
not in info:  patman
not in info:  jiererniure
not in info:  Konop1992
not in info:  pavelm
not in info:  pavelm
not in info:  pavelm
not in info:  pavelm
not in info:  hawkeye9797-2
not in info:  Messer
not in info:  asifmallik
not in info:  gokturkgoker
not in info:  jolvr
not in info:  Nakel
not in info:  3cucilo
not in info:  erashid
not in info:  erashid
not in info:  mburch1974
not in info:  Lasin
not in info:  Lasin
not in info:  alik
not in info:  alik
not in info:  Faz
not in info:  Faz
not in info:  Krylo
not in info:  Krylo
not in info:  darkbit
not in info:  darkbit
not in info:  aleksi
not in info:  darkbit
not in info:  darkbit
not in info:  Fosil
not in info:  Fosil
not in info:  Fosil
not in info:  Fosil
not in info:  chessprocator
not in info:  bravone
not in info:  lukasz1983
not in info:  FMB
not in info:  leszek_75-75
not in info:  leszek_75-75
not in info:  leszek_75-75
not in info:  amithebond
not in info:  amithebond
not in info:  Nitra6
not in info:  artbrokes
not in info:  absimaldata
not in info:  absimaldata
not in info:  absimaldata
not in info:  NachoSoto
Exception on run:  5144513ee4721c1c36a06a58
Successfully updated 1329 users
You have new mail in /var/mail/usr00
```

**Comment by ppigazzini** [comment]
@tomtor PROD updated, I will start a scan all on PROD in 15 minutes (after the hourly backup).

**Comment by vondele** [comment]
congrats guys!

**Comment by ppigazzini** [comment]
@tomtor scan_all finished successfully on PROD, crontab restored.

**Comment by ppigazzini** [comment]
@tomtor IIRC pserve on python 3 is using half RAM wrt on python 2

**Comment by tomtor** [comment]
@ppigazzini I squashed the commits into 3 topics. Ok to merge for me.

**Comment by ppigazzini** [comment]
PROD server update, thank you @tomtor :)

**Pull request merged**